src/psyclone/psyir/transformations/inline_trans.py

```diff
@@ -0,0 +1,406 @@
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: A. R. Porter, STFC Daresbury Lab
+
+'''
+This module contains the InlineTrans transformation.
+
+'''
+
+from psyclone.errors import InternalError
+from psyclone.psyGen import Transformation
+from psyclone.psyir.nodes import (
+    Call, Routine, Reference, CodeBlock, Return, Literal)
+from psyclone.psyir.symbols import (ContainerSymbol, DataSymbol, ScalarType,
+                                    ImportInterface)
+from psyclone.psyir.transformations.transformation_error import (
+    TransformationError)
+
+
+class InlineTrans(Transformation):
+    '''
+    This transformation takes a Call and replaces it with the body of the
+    target routine. It is used as follows:
+
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.nodes import Call, Routine
+    >>> from psyclone.psyir.transformations import InlineTrans
+    >>> code = ("""
+    ... module test_mod
+    ... contains
+    ...   subroutine run_it()
+    ...   integer :: i
+    ...   real :: a(10)
+    ...   do i=1,10
+    ...     a(i) = 1.0
+    ...     call sub(a(i))
+    ...   end do
+    ...   end subroutine run_it
+    ...   subroutine sub(x)
+    ...     real, intent(inout) :: x
+    ...     x = 2.0*x
+    ...   end subroutine sub
+    ... end module test_mod"""
+    >>> psyir = FortranReader().psyir_from_source(code)
+    >>> call = psyir.walk(Call)[0]
+    >>> inline_trans = InlineTrans()
+    >>> inline_trans.apply(call)
+    >>> # Uncomment the following line to see a text view of the schedule
+    >>> # print(psyir.walk(Routine)[0].view())
```

Comment:
Also, if possible I prefer finishing the doctest with e.g. `>>> print(FortranWriter()(psyir.walk(Routine)[0]))` and then have the output displayed (e.g like in src/psyclone/psyir/transformations/hoist_trans.py)

It makes for clear documentation and a complete doctest check

Reply:
Good idea and thanks for pointing me at an example.